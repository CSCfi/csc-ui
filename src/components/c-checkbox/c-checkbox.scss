:host {
  display: block;
}

.c-checkbox {
  $this: &;

  --c-checkbox-color: var(--csc-primary);

  align-items: center;
  display: flex;
  position: relative;
  cursor: pointer;
  font-size: 16px;
  user-select: none;
  gap: 4px;

  input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;

    &:checked ~ .ripple .checkmark {
      box-shadow: inset 0 0 0 20px var(--c-checkbox-color);

      &::after {
        display: block;
      }
    }
  }

  .ripple {
    position: relative;
    height: 42px;
    width: 42px;
    border-radius: 50%;
    overflow: hidden;
    transition: translateZ(0);
    color: var(--csc-primary);
  }

  .checkmark {
    position: absolute;
    top: 12px;
    left: 12px;
    height: 18px;
    width: 18px;
    background-color: transparent;
    box-shadow: inset 0 0 0 2px var(--c-checkbox-color);
    border-radius: 2px;
    transition: box-shadow 0.15s ease-in-out;

    &::after {
      content: '';
      position: absolute;
      display: none;
      left: 6px;
      top: 1px;
      width: 6px;
      height: 12px;
      border: solid white;
      border-width: 0 2px 2px 0;
      transform: rotate(45deg);
    }
  }

  &__details {
    line-height: 1;
    padding: 0 12px;

    &.active {
      #{$this} {
        &__message {
          opacity: 1;
          transform: translateY(0px);
        }
      }
    }
  }

  &__message {
    font-size: 12px;
    opacity: 0;
    transform: translateY(-4px);
    transition: all 0.2s cubic-bezier(0.25, 0.8, 0.5, 1);

    &--hint {
      color: var(--csc-mid-grey);
    }

    &--error {
      color: var(--csc-error);
    }

    span {
      align-items: center;
      color: currentColor;
      display: flex;
      gap: 4px;
      height: 16px;
    }

    svg {
      fill: currentColor;
    }
  }

  &--disabled {
    --c-checkbox-color: var(--csc-mid-grey);

    color: var(--c-checkbox-color);
    cursor: default;
    opacity: 0.75;
  }

  &:focus {
    outline: none;
  }

  &:focus .ripple {
    outline: 2px var(--c-checkbox-color) solid;
  }

  @supports selector(:focus-visible) {
    &:focus .ripple {
      outline: none;
    }
  }

  &:focus-visible .ripple {
    outline: 2px var(--c-checkbox-color) solid;
  }

  &:not(.c-checkbox--disabled):hover {
    .ripple {
      background-color: var(--csc-primary-text-hover);
    }
  }
}
