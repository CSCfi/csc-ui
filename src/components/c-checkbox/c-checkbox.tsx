import {
  Component,
  Host,
  h,
  Listen,
  Prop,
  Event,
  EventEmitter,
} from '@stencil/core';
import { createRipple } from '../../utils/utils';
/**
 * @group Form
 */
@Component({
  tag: 'c-checkbox',
  styleUrl: 'c-checkbox.scss',
  shadow: true,
})
export class CCheckbox {
  /**
   * Element label
   */
  @Prop() label: string = '';

  /**
   * Is the element checked
   */
  @Prop({ mutable: true }) value: boolean = false;

  /**
   * Disable the checkbox
   */
  @Prop() disabled = false;

  /**
   * Triggered when element is checked or unchecked
   */
  @Event() changeValue: EventEmitter;

  private _container: HTMLDivElement;

  @Listen('keydown', { passive: true })
  handleKeyDown(event: KeyboardEvent) {
    if (['Space', 'Enter'].includes(event.code)) {
      event.preventDefault();
      this.toggleState(event);
    }
  }

  private toggleState(event) {
    if (this.disabled) return;

    createRipple(event, this._container, true);

    this.value = !this.value;
    this.changeValue.emit(this.value);
  }

  render() {
    const wrapperClasses = {
      'c-checkbox': true,
      'c-checkbox--disabled': this.disabled,
    };

    return (
      <Host>
        <label
          id="c-checkbox-label"
          class={wrapperClasses}
          tabindex={this.disabled ? -1 : 0}
        >
          <input
            type="checkbox"
            aria-checked={this.value}
            aria-disabled={this.disabled}
            aria-labelledby="c-checkbox-label"
            checked={this.value}
            disabled={this.disabled}
            tabindex="-1"
            onChange={(event) => this.toggleState(event)}
          />
          <div
            class="ripple"
            ref={(el) => (this._container = el as HTMLDivElement)}
          >
            <span class="checkmark"></span>
          </div>
          {this.label}
        </label>
      </Host>
    );
  }
}
