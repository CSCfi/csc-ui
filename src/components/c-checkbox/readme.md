# c-checkbox



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description            | Type      | Default |
| -------- | --------- | ---------------------- | --------- | ------- |
| `label`  | `label`   | Element label          | `string`  | `''`    |
| `value`  | `value`   | Is the element checked | `boolean` | `false` |


## Events

| Event         | Description                                    | Type               |
| ------------- | ---------------------------------------------- | ------------------ |
| `changeValue` | Triggered when element is checked or unchecked | `CustomEvent<any>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
