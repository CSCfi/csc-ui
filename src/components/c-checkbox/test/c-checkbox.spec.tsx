import { newSpecPage } from '@stencil/core/testing';
import { CCheckbox } from '../c-checkbox';

describe('c-checkbox', () => {
  it('renders default html', async () => {
    const page = await newSpecPage({
      components: [CCheckbox],
      html: `<c-checkbox></c-checkbox>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
