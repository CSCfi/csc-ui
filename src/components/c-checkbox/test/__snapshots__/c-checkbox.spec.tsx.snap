// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-checkbox renders default html 1`] = `
<c-checkbox>
  <mock:shadow-root>
    <div class="c-checkbox">
      <input aria-checked="false" aria-disabled="false" class="visuallyhidden" id="checkbox" type="checkbox">
      <label class="c-checkbox__label" htmlfor="checkbox">
        <div class="ripple">
          <svg viewBox="0 0 100 100"></svg>
        </div>
        <div class="c-checkbox__label-content">
          <slot></slot>
        </div>
      </label>
    </div>
    <div class="c-checkbox__details">
      <div class="c-checkbox__message c-checkbox__message--hint"></div>
    </div>
  </mock:shadow-root>
</c-checkbox>
`;
